File: vuln-sandbox/typescript/cwe-22/path-traversal.ts
// VULNERABLE: InputPathTraversal — path traversal via user-controlled filename (TypeScript)
// Rule: InputPathTraversal | CWE-22 | Severity: HIGH

import express, { Request, Response } from 'express';
import fs from 'fs';
import path from 'path';

const app = express();
const UPLOAD_DIR = '/var/app/uploads';

app.get('/file', (req: Request, res: Response): void => {
  const filename: string = req.query.file as string;

  // VULNERABLE: user-controlled filename used to construct file path
  // An attacker can pass: ../../etc/passwd to read arbitrary files
  const filePath: string = path.join(UPLOAD_DIR, filename);

  fs.readFile(filePath, (err, data) => {
    if (err) {
      res.status(404).json({ error: 'File not found' });
      return;
    }
    res.send(data);
  });
});

app.listen(3000);

File: vuln-sandbox/rust/cwe-22/path_traversal.rs
// VULNERABLE: rust-fs-read-format — std::fs::read with format! path
// Rule: rust-fs-read-format | CWE-22 | Severity: HIGH

use std::fs;

fn read_user_file(user_filename: &str) -> Vec<u8> {
    // VULNERABLE: user-controlled filename interpolated into file path via format!
    // An attacker can pass: ../../etc/passwd to read arbitrary files
    let data = std::fs::read(format!("/uploads/{}", user_filename))
        .unwrap_or_default();
    data
}

fn main() {
    // Simulating user input from an HTTP request parameter
    let filename = std::env::args().nth(1).unwrap_or_default();
    let content = read_user_file(&filename);
    println!("read {} bytes", content.len());
}
